engine/runtime-language-arrow/src/main/java/org/enso/interpreter/arrow/runtime/ArrowFixedSizeArrayBuilder.java

```diff
@@ -65,33 +68,83 @@ Object getMembers(boolean includeInternal) throws UnsupportedMessageException {
   }
 
   @ExportMessage
-  Object invokeMember(
-      String name,
-      Object[] args,
-      @Cached(value = "buildWriterOrNull(name)", neverDefault = true)
-          WriteToBuilderNode writeToBuilderNode)
+  Object invokeMember(String name, Object[] args, @Cached AppendNode append)
       throws UnsupportedMessageException, UnknownIdentifierException, UnsupportedTypeException {
-    switch (name) {
-      case BUILD_OP:
-        sealed = true;
-        return switch (unit) {
-          case Date32, Date64 -> new ArrowFixedArrayDate(buffer, size, unit);
-          case Int8, Int16, Int32, Int64 -> new ArrowFixedArrayInt(buffer, size, unit);
-        };
-      case APPEND_OP:
-        if (sealed) {
-          throw UnsupportedMessageException.create();
-        }
-        var current = index;
-        writeToBuilderNode.executeWrite(this, current, args[0]);
-        index += 1;
-        return NullValue.get();
-      default:
-        throw UnknownIdentifierException.create(name);
+    return switch (name) {
+      case BUILD_OP -> build();
+      case APPEND_OP -> {
+        append.executeAppend(this, args[0]);
+        yield NullValue.get();
+      }
+      default -> throw UnknownIdentifierException.create(name);
+    };
+  }
+
+  private final TruffleObject build() throws UnsupportedMessageException {
+    var b = buffer;
+    if (b == null) {
+      throw UnsupportedMessageException.create();
+    }
+    buffer = null;
+    return switch (unit) {
+      case Date32, Date64 -> new ArrowFixedArrayDate(b, size, unit);
+      case Int8, Int16, Int32, Int64 -> new ArrowFixedArrayInt(b, size, unit);
+    };
+  }
+
+  @GenerateUncached
+  @GenerateInline(false)
+  abstract static class AppendNode extends Node {
+    abstract void executeAppend(ArrowFixedSizeArrayBuilder builder, Object value)
+        throws UnsupportedTypeException, UnsupportedMessageException;
+
+    @Specialization(
+        limit = "3",
+        guards = {"builder.getUnit() == cachedUnit"})
+    static void writeToBuffer(
+        ArrowFixedSizeArrayBuilder builder,
+        Object value,
+        @Cached(value = "builder.getUnit()", allowUncached = true) LogicalLayout cachedUnit,
+        @Shared("put") @Cached ByteBufferDirect.PutNode put,
+        @Shared("value") @Cached ValueToNumberNode valueNode,
+        @Shared("iop") @CachedLibrary(limit = "3") InteropLibrary iop)
+        throws UnsupportedTypeException, UnsupportedMessageException {
+      if (iop.isNull(value)) {
+        put.putNull(builder.buffer, cachedUnit);
+        return;
+      }
+      var number = valueNode.executeAdjust(cachedUnit, value);
```

Comment:
Unlike the previous version that was [converting and calling putXyz at once place](https://github.com/enso-org/enso/pull/10150/files#diff-fa3a6b1cc4238b0d6fa4e7faa0f2a1197a4b7f13596308056f584e31af430dd5L52) the `executeAdjust` just converts `value` to appropriate `java.lang.Number` and then we use the `PutNode` to place that value into the buffer.

Truffle gives us a way to separate concerns for operations without loosing the speed - the whole `switch` will be _compiled away_ and at the end will look just like the old one, but in source it is more structured into individual nodes.